src/dialog-composition-engine.ts

```diff
@@ -0,0 +1,137 @@
+import { CompositionContext, CompositionEngine, ViewSlot } from 'aurelia-templating';
+import { DialogSettings } from './dialog-settings';
+import { invokeLifecycle } from './lifecycle';
+import { createDialogCancelError } from './dialog-cancel-error';
+import { createDialogCloseError } from './dialog-close-error';
+import { DialogOpenResult } from './dialog-result';
+import { DialogController } from './dialog-controller';
+import { InfrastructureDialogController } from './infrastructure-dialog-controller';
+
+interface DialogCompositionInstruction {
+    context: CompositionContext;
+    controller: InfrastructureDialogController;
+}
+
+function canActivate(instruction: DialogCompositionInstruction): Promise<DialogCompositionInstruction> {
+    if (!instruction.context.viewModel) { return Promise.resolve(instruction); }
+    return invokeLifecycle(instruction.context.viewModel, 'canActivate', instruction.controller.settings.model)
+        .then(canActivate => {
+            if (canActivate) { return instruction; }
+            throw createDialogCancelError();
+        });
+}
+
+// tslint:disable:no-empty
+const noopViewSlot: ViewSlot = {
+    animateView() { },
+    add() { },
+    insert() { },
+    move() { },
+    remove() { },
+    removeAt() { return {} as any; },
+    removeMany() { },
+    removeAll() { },
+    projectTo() { },
+    bind() { },
+    attached() { },
+    detached() { },
+    unbind() { },
+    transformChildNodesIntoView() { }
+};
+// tslint:enable:no-empty
+(noopViewSlot as any).children = [];
+
+/**
+ * @internal
+ */
+export class DialogCompositionEngine {
+    public static inject = [Container, CompositionEngine];
+    constructor(
+        private container: Container,
+        private compositionEngine: CompositionEngine
+    ) { }
+
+    public createDialogController(container: Container, settings: DialogSettings): InfrastructureDialogController {
+        const controller = container.invoke(DialogController, [settings]);
+        container.registerInstance(DialogController, controller);
+        return controller;
+    }
+
+    public createCompositionContext(
+        childContainer: Container,
+        controller: InfrastructureDialogController
+    ): CompositionContext {
+        const context: CompositionContext = {
+            container: childContainer.parent,
+            childContainer,
+            bindingContext: null,
+            overrideContext: null,
+            viewResources: null as any,
+            model: controller.settings.model,
+            view: controller.settings.view,
+            viewModel: controller.settings.viewModel,
+            // we don't want the CompositionEngine to add the View to a ViewSlot,
+            // that's the job of the Renderer
+            viewSlot: noopViewSlot,
+            host: controller.renderer.getDialogContainer()
+        };
+        if (typeof context.viewModel === 'function') {
+            const moduleId = Origin.get(context.viewModel).moduleId;
```

Comment:
Isn't this needed only if we don't have a view?

Reply:
@EisenbergEffect no, this is the case where you pass a `class` for the `viewModel`.
```
import {MyDialogViewModel} from './my-dialog';
//...
dialogService.open({viewModel: MyDialogViewModel})
//...
```
And this is the case that has the *greatest* issues with `WebPack` and needs a bit more work from the one using it to make it work.
This is supported by the dialog from the early `beta`s.

Reply:
Cool. Just checking. It's been a while since I looked at this ;)